### PR TITLE
Make BIGNUM rand functions available within the FIPS module

### PR DESCRIPTION
**Pull request by mattcaswell**
PR #9130 made the BIGNUM code available from within the FIPS provider, but had the RAND functions disabled because no RAND code was available from inside the provider at that time. When PR #9035 is approved, RAND will become available from within the provider.

This PR re-enables the BIGNUM RAND code. It depends on #9035 so is marked as WIP for now.

### CODE REVIEW

**Comment by levitte** [comment]
https://travis-ci.org/openssl/openssl/jobs/547758528#L2447 looks relevant...  https://ci.appveyor.com/project/openssl/openssl/builds/25392229/job/gvnaly2ws8q8u21c#L5014 says the same thing...

**Comment by mattcaswell** [comment]
Now that #9035 has gone in I've rebased this and taken it out of WIP.

Needs a reviewer...

File: crypto/bn/bn_rand.c [hunk 1]
```diff
@@ -111,23 +99,33 @@ static int bnrand(BNRAND_FLAG flag, BIGNUM *rnd, int bits, int top, int bottom)
...
 }
 
+int BN_rand_ex(BIGNUM *rnd, int bits, int top, int bottom, BN_CTX *ctx)
+{
+    return bnrand(NORMAL, rnd, bits, top, bottom, ctx);
+}
 int BN_rand(BIGNUM *rnd, int bits, int top, int bottom)
 {
-    return bnrand(NORMAL, rnd, bits, top, bottom);
+    return BN_rand_ex(rnd, bits, top, bottom, NULL);
```
Comment (mspncp) [inline crypto/bn/bn_rand.c]: Since the actual implementation is in the private helper function named `bnrand()`, I would prefer if both public implementations `BN_rand()` and `BN_rand_ex()` would be based on it, instead of basing the implementation of the former on the latter

```C
int BN_rand(BIGNUM *rnd, int bits, int top, int bottom)
{
   return bnrand(NORMAL, rnd, bits, top, bottom, NULL);
}
```

File: crypto/bn/bn_rand.c [hunk 2]
```diff
@@ -111,23 +99,33 @@ static int bnrand(BNRAND_FLAG flag, BIGNUM *rnd, int bits, int top, int bottom)
...
+
+int BN_priv_rand_ex(BIGNUM *rnd, int bits, int top, int bottom, BN_CTX *ctx)
+{
+    return bnrand(PRIVATE, rnd, bits, top, bottom, ctx);
 }
 
 int BN_priv_rand(BIGNUM *rnd, int bits, int top, int bottom)
 {
-    return bnrand(PRIVATE, rnd, bits, top, bottom);
+    return BN_priv_rand_ex(rnd, bits, top, bottom, NULL);
```
Comment (mspncp) [inline crypto/bn/bn_rand.c]: dito

File: crypto/bn/bn_rand.c [hunk 3]
```diff
@@ -191,14 +190,24 @@ static int bnrand_range(BNRAND_FLAG flag, BIGNUM *r, const BIGNUM *range)
...
 
+int BN_rand_range_ex(BIGNUM *r, const BIGNUM *range, BN_CTX *ctx)
+{
+    return bnrand_range(NORMAL, r, range, ctx);
+}
+
 int BN_rand_range(BIGNUM *r, const BIGNUM *range)
 {
-    return bnrand_range(NORMAL, r, range);
+    return BN_rand_range_ex(r, range, NULL);
```
Comment (mspncp) [inline crypto/bn/bn_rand.c]: dito

File: crypto/bn/bn_rand.c [hunk 4]
```diff
@@ -191,14 +190,24 @@ static int bnrand_range(BNRAND_FLAG flag, BIGNUM *r, const BIGNUM *range)
...
+
+int BN_priv_rand_range_ex(BIGNUM *r, const BIGNUM *range, BN_CTX *ctx)
+{
+    return bnrand_range(PRIVATE, r, range, ctx);
 }
 
 int BN_priv_rand_range(BIGNUM *r, const BIGNUM *range)
 {
-    return bnrand_range(PRIVATE, r, range);
+    return BN_priv_rand_range_ex(r, range, NULL);
```
Comment (mspncp) [inline crypto/bn/bn_rand.c]: dito

File: crypto/rand/rand_lib.c [hunk 5]
```diff
@@ -749,31 +749,52 @@ void RAND_add(const void *buf, int num, double randomness)
  * the default method, then just call RAND_bytes().  Otherwise make
  * sure we're instantiated and use the private DRBG.
  */
-int RAND_priv_bytes(unsigned char *buf, int num)
+int rand_priv_bytes_ex(OPENSSL_CTX *ctx, unsigned char *buf, int num)
```
Comment (mspncp) [inline crypto/rand/rand_lib.c]: IMHO the addition of the _ex functions in rand_lib.c should be a separate commit preceding this commit.

**Comment by mattcaswell** [comment]
I've made the requested changes, and split the first commit into two as suggested. Due to the changes in the number of commits, I've force pushed this.

**Comment by mattcaswell** [comment]
Rebased to resolve a conflict with master.

Ping?

File: doc/internal/man3/rand_bytes_ex.pod [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+ int rand_bytes_ex(OPENSSL_CTX *ctx, unsigned char *buf, int num);
+ int rand_priv_bytes_ex(OPENSSL_CTX *ctx, unsigned char *buf, int num);
+
+=head1 DESCRIPTION
+
+rand_bytes_ex and rand_priv_bytes_ex are the equivalent of RAND_bytes() and
+RAND_priv_bytes() in the public API except that they both take an additional
+B<ctx> parameter.
+The DRBG used for the operation is the public or private DRBG as appropriate
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply doc/internal/man3/rand_bytes_ex.pod]: Its an English idiom (British English?) meaning "as appropriate to the situation", i.e. in this context we use the public DRBG in rand_bytes_ex and the private DRBG in rand_priv_bytes_ex.

https://www.macmillandictionary.com/dictionary/british/as-appropriate

On re-reading it, it still makes sense to me, but I deleted it anyway.

File: doc/internal/man3/rand_bytes_ex.pod [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+associated with the specified B<ctx>. The parameter can be NULL in which case
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply doc/internal/man3/rand_bytes_ex.pod]: Added.

File: doc/internal/man3/rand_bytes_ex.pod [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+the default library ctx is used. If the default RAND_METHOD has been changed
+then this overrides any library ctx and the RAND_METHOD implementation will be
+used in preference.
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply doc/internal/man3/rand_bytes_ex.pod]: Done

File: doc/internal/man3/rand_bytes_ex.pod [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+=head1 RETURN VALUES
+
+rand_bytes_ex and rand_bytes_priv_ex return 0 or less on error or 1 on success.
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply doc/internal/man3/rand_bytes_ex.pod]: Fixed.

**Comment by mattcaswell** [comment]
Fixup commit pushed addressing nits.

@mspncp can you reconfirm?

**Comment by mattcaswell** [comment]
Pushed. Thanks.
